target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch

```diff
@@ -0,0 +1,86 @@
+diff -Naurp a/arch/mips/bcm63xx/boards/board_bcm963xx.c b/arch/mips/bcm63xx/boards/board_bcm963xx.c
+--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c	2018-02-19 04:18:41.127920388 +0100
++++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c	2018-02-19 04:17:04.958906434 +0100
+@@ -2193,6 +2193,66 @@ static struct board_info __initdata boar
+ 	},
+ };
+ 
++static struct sprom_fixup __initdata vr3022eu_fixups[] = {
++	{ .offset = 97, .value = 0xfeb3 },
++	{ .offset = 98, .value = 0x1618 },
++	{ .offset = 99, .value = 0xfab0 },
++	{ .offset = 113, .value = 0xfed1 },
++	{ .offset = 114, .value = 0x1609 },
++	{ .offset = 115, .value = 0xfad9 },
++};
```

Comment:
I copied these from the `vr3025u_fixups`. I don't know how to asses whether it's correct or not. The BCM43222 seems to work though.

Reply:
You should find these in the appropriate GPL tarball. There should be a 43222_map.bin or so for the values.

Reply:
There are the following files in the GPL tarball (which may not the tarball for the exact router model, but probably close):
/bcmdrivers/broadcom/net/wl/impl8/bcm4322_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43222_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43224_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43225_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43226_map.bin_save

In addition to that, there are these files in the original firmware's filesystem:
/etc/wlan/bcm4322_map.bin
/etc/wlan/bcm43222_map.bin
/etc/wlan/bcm43224_map.bin
/etc/wlan/bcm43225_map.bin
/etc/wlan/bcm43226_map.bin
/etc/wlan/bcm43227_map.bin
/etc/wlan/bcm43228_map.bin

The corresponding files from these two sources are the same except for bcm43225_map.bin_save/bcm43225_map.bin which differ in 12 bytes.

Which file do I need and how do I use it to verify the fixups are correct.

Reply:
I found the templates in `arch/mips/bcm63xx/sprom.c`. It turns out the `/etc/wlan/bcm43222_map.bin` is quite similar to the `bcm43222_sprom` template, except the very end of the file:

* /etc/wlan/bcm43222_map.bin:
  ```
  ...
  0001a0 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
  0001b0 ff ff ff ff ff ff 44 08 00 00 00 00 00 00 00 00
  ```
* bcm43222_sprom:
  ```
  ...
  0001a0 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
  0001b0 ff ff ff ff ff ff 00 08
  ```
The `/etc/wlan/bcm43222_map.bin` is 8 bytes longer! (All those extra bytes are zeroes though.)
And there is a difference in the last but one byte of the `bcm43222_sprom` (0x44 vs. 0x00).

I'm wondering could that be some kind of a checksum or does that rather mean I need the following fixup:
```c
static struct sprom_fixup __initdata vr3022eu_fixups[] = {
        { .offset = 219, .value = 0x4408 },
};
```

Reply:
I noticed the `EVG2000_fixups` from `arch/mips/bcm63xx/boards/board_bcm963xx.c` is quite similar to what I'm seeing. So hopefully the fixup I arrived at above makes sense.